Alignment with Jakarta EE 9

Emily-Jiang opened the pull request:
Fixes: #668

radcortez commented on api/src/main/java/org/eclipse/microprofile/config/inject/package-info.java:
```diff
@@ -44,5 +44,5 @@
  * @author <a href="mailto:emijiang@uk.ibm.com">Emily Jiang</a>
  */
-@org.osgi.annotation.versioning.Version("1.1")
+@org.osgi.annotation.versioning.Version("2.0")
```
How about if we align these versions?

Emily-Jiang replied:
This is the semantic version for the corresponding package. They are all individually version to indicate whether there are any backward incompatible changes between teh releases. There should not be any alignment with the different packages. As you might have noticed, not all of the packages have changed their versions. For more [info](http://docs.osgi.org/whitepaper/semantic-versioning/040-semantic-versions.html) regarding semantic versioning. We use bnd versioning plugin to work out all of the versions.

radcortez replied:
I know, but this is almost like starting fresh, and I think it wouldn't hurt to align all versions (even if there are no changes, but to mark a baseline with the Jakarta rename) and then follow the regular versioning scheme going forward.

Emily-Jiang replied:
It's not like starting from scratch though. We have been follow the regular versioning scheme from the beginning. I think it is expecially important to obey the rules as you can see some packages are not even affected by the changes so their version will not be updated. If we directly jump from 1.1->3.0, this will confuse end users.

Emily-Jiang replied:
@radcortez do you have further comments? If not, can you approve this PR so that I can make further progress. At the moment, I am blocked by this.

bjhargrave replied:
Package versions should be semantic. Bundle versions and specification document versions are free to change. 

If you change the major version of a package then you affect all consumers of the package by breaking their imports as they, properly, import to exclude a major version bump as that means a breaking change.

Each package has an independent semantic versioning life. There is no reason to keep all package versions in lock step and there is no reason to change a major version unless there is a breaking change you need to communicate to users of the package.

A package's version is a message to the future. Make sure the message is clear and not noise.

radcortez commented on api/src/main/java/org/eclipse/microprofile/config/package-info.java:
```diff
@@ -82,5 +82,5 @@
  * @author <a href="mailto:struberg@apache.org">Mark Struberg</a>
  */
-@org.osgi.annotation.versioning.Version("2.0.0")
+@org.osgi.annotation.versioning.Version("3.0.0")
```
How about if we align these versions?

Emily-Jiang replied:
@radcortez please see my comments [here](https://github.com/eclipse/microprofile-config/pull/699#discussion_r640165314)

radcortez reviewed (commented):
Don't we need changes in the spec?

Emily-Jiang commented on api/src/main/java/org/eclipse/microprofile/config/spi/package-info.java:
```diff
@@ -32,5 +32,5 @@
  *
  */
-@org.osgi.annotation.versioning.Version("2.0")
+@org.osgi.annotation.versioning.Version("3.0")
```
same as above.

ederks85 reviewed (approved):
LGTM. I have zero knowledge that might help the OSGI alignment discussion., unfortunately.

Emily-Jiang commented:
> LGTM. I have zero knowledge that might help the OSGI alignment discussion., unfortunately.

Thank you @ederks85 ! MP uses OSGi's semantic versioning policy and its plugin to work out backward incompatible changes. MP has benefited from this tool along the way to spell out backward incompatible changes and version the releases appropriately.

radcortez commented:
Feel free to do as you wish. I still we could align all version (it is ok to jump versions) and it would be easier to maintain in the future, then keeping track different versions for all the separate packages (and since we are breaking the API is ok to push new versions).

Emily-Jiang commented:
> Feel free to do as you wish. I still we could align all version (it is ok to jump versions) and it would be easier to maintain in the future, then keeping track different versions for all the separate packages (and since we are breaking the API is ok to push new versions).

There is difference between api package version and spec version. For instance, one package has not changed since the first release of MP Config, there is no reason to randomly bump up its version if no change was made since then. In this way, end users will immediately find out that this package has not got any changes since the previous releases. This is what semantic versioning is about. For spec version, if there is backward incompatible changes in any packages, the spec version will need to be bumped to indicate backward incompatible changes are in the current release.

radcortez commented:
We are breaking the package and we are bumping the versions it not like we are randomly bumping the version. My suggestion was to just align all versions (not necessarily align with the project version).

Emily-Jiang commented:
> We are breaking the package and we are bumping the versions it not like we are randomly bumping the version. My suggestion was to just align all versions (not necessarily align with the project version).

@radcortez I appreciate your thoughts on this. For the packages that do not rely on any jakarta namespace, we should just leave them alone. We have smenatic versioning plugin to help us, we can do it precisely correctly. Therefore, I am against the idea of bumping all packages unnecessarily.
